Question 3
The question is about causal inference with a continuous response y, 0-1 valued treatment t, and
confounders x = (x0, x1, x2, x3). The sample size n = 2000 and training data (yi, ti, xi) (i =
1, . . . , n) are in files resp.csv, trt.csv, and conf.csv.
1. (10 pts) Estimate the average treatment effect (ATE) using propensity scores. Employ logistic
regression to estimate the propensity scores necessary for computing the ATE.

In [28]:
import pandas as pd
import numpy as np

# Load the data
y = np.array(pd.read_csv('resp.csv'))
t = np.array(pd.read_csv('trt.csv'))
X = np.array(pd.read_csv('conf.csv'))

In [23]:
np.array(y)

array([[-0.27178862],
       [-0.19972188],
       [ 2.66125352],
       ...,
       [ 3.80922299],
       [-0.60139812],
       [ 4.23244633]])

In [29]:
from sklearn.linear_model import LogisticRegression

# Fit logistic regression
logit = LogisticRegression(solver='lbfgs')
logit.fit(X, t)

# Estimate propensity scores
p_score = logit.predict_proba(X)[:, 1]

# Estimate ATE using propensity scores
ate_logit = np.mean((t * y) / p_score) - np.mean(((1 - t) * y) / (1 - p_score))
print(f"ATE using logistic regression propensity scores: {ate_logit:.4f}")

ATE using logistic regression propensity scores: -1.5174


c:\Users\PERFECT SOLUTION\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\utils\validation.py:1300: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [35]:
logit.predict_proba(X)[:,1]

array([0.37769339, 0.88302675, 0.24605851, ..., 0.15517218, 0.68493871,
       0.22510504])

2. (10 pts) Instead of logistic regression, use a deep neural network to estimate the propensity scores,
and then compute the ATE. Compare the ATE estimate with the previous ATE estimate.

In [36]:
from sklearn.neural_network import MLPClassifier

# Fit deep neural network
nn = MLPClassifier(hidden_layer_sizes=(64, 32), max_iter=500, early_stopping=True)
nn.fit(X, t)

# Estimate propensity scores
p_score_nn = nn.predict_proba(X)[:, 1]

# Estimate ATE using propensity scores
ate_nn = np.mean((t * y) / p_score_nn) - np.mean(((1 - t) * y) / (1 - p_score_nn))
print(f"ATE using deep neural network propensity scores: {ate_nn:.4f}")

c:\Users\PERFECT SOLUTION\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1101: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


ATE using deep neural network propensity scores: -0.7958


A negative ATE value suggests that, on average, the treatment has a negative effect on the outcome variable compared to the control group (no treatment).
Here's an interpretation of these results:

ATE using logistic regression propensity scores: -1.5174

The estimated ATE using logistic regression propensity scores is -1.5174.
This means that, on average, the treatment reduces the outcome variable by approximately 1.5174 units compared to the control group, after accounting for the confounding variables (x0, x1, x2, x3).
A negative ATE implies that the treatment has a harmful or undesirable effect on the outcome variable.


ATE using deep neural network propensity scores: -0.7958

The estimated ATE using the deep neural network propensity scores is -0.7958.
This means that, on average, the treatment reduces the outcome variable by approximately 0.7958 units compared to the control group, after accounting for the confounding variables (x0, x1, x2, x3).
The magnitude of the negative ATE is smaller compared to the logistic regression approach, suggesting a less severe negative effect of the treatment on the outcome variable.

Comparing the two ATE estimates:

Both methods suggest a negative effect of the treatment on the outcome variable, but the deep neural network approach estimates a smaller negative effect (-0.7958) compared to the logistic regression approach (-1.5174).

Deep neural networks can potentially model more complex non-linear relationships, which might lead to a better estimation of the propensity scores and, consequently, a different ATE estimate.